Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift

```diff
@@ -6,102 +6,167 @@
 
 import Foundation
 
-// TODO: RUMM-300 plug DDTracer into swizzled methods
-
-internal class URLSessionSwizzler {
-    enum Selectors {
-        static let DataTaskWithURL = #selector(URLSession.dataTask(with:) as (URLSession) -> (URL) -> URLSessionDataTask)
-        static let DataTaskWithRequest = #selector(URLSession.dataTask(with:) as (URLSession) -> (URLRequest) -> URLSessionDataTask)
-        static let DataTaskWithURLCompletion = #selector(URLSession.dataTask(with:completionHandler:) as (URLSession) -> (URL, @escaping CompletionHandler) -> URLSessionDataTask)
-        static let DataTaskWithRequestCompletion = #selector(URLSession.dataTask(with:completionHandler:) as (URLSession) -> (URLRequest, @escaping CompletionHandler) -> URLSessionDataTask)
+// hook into URLSessionTask creation
+internal typealias RequestInterceptor = (URLRequest) -> InterceptionResult?
+
+// modifying URLRequest for task creation, observer block to run at resume/completion
+internal typealias InterceptionResult = (request: URLRequest, taskPayload: TaskObserver)
+internal typealias TaskObserver = (TaskObservationEvent) -> Void
+internal enum TaskObservationEvent: Int, Equatable {
+    case starting
+    case completed
+}
+
+internal func swizzleURLSession(interceptor: @escaping RequestInterceptor) throws {
+    do {
+        try URLSessionSwizzler.DataTask_URL_Completion.swizzle(interceptor: interceptor)
+        try URLSessionSwizzler.DataTask_Request_Completion.swizzle(interceptor: interceptor)
+    } catch {
+        try URLSessionSwizzler.DataTask_URL_Completion.unswizzle()
+        try URLSessionSwizzler.DataTask_Request_Completion.unswizzle()
+        throw error
```

Comment:
How will we handle the `unswizzle()` errors? It seems that we cannot guarantee a non-intrusive behaviour for the customer app with the current implementation.

Reply:
you are right 👍 

it should be
```swift
catch {
  try? URLSessionSwizzler.DataTask_URL_Completion.unswizzle()
  try? URLSessionSwizzler.DataTask_Request_Completion.unswizzle()
  throw error
}
```
as we attempt to revert, but if that fails too we can't do anything further and we should give the user what went wrong in the first place

Reply:
But `try?` will leave the customer app in the inconsistent state as it only ignores the failure ⚠️.

The problem is that we only know if the swizzling error happens while we're swizzling, which already modifies the things:
```swift
try URLSessionSwizzler.DataTask_URL_Completion.swizzle(interceptor: interceptor)
```

A typical approach to this is to separate configuration from actual logic:
```swift
// instead of:
do {
   try doTheThing()
} catch {
   // 🤷‍♂️ can't undo, the thing is PARTIALLY done
}

// make it:
do {
   let preparation = try prepareTheThing()
   doTheThing(preparation)
} catch {
   // 👍 tell the user we can't do the thing
   // ⭐️ nothing to undo as the thing was not done
}
```

Reply:
looks like a good idea 👍 
i made the change but that led to a non-trivial refactoring, you may wanna review this PR once again

`MethodSwizzler` was throwing only in the case of "method not found".
I changed its API in a way that methods take `FoundMethod`, that way methods become non-throwing.
Now we first find the method, only if method is found then we go on with swizzling; so no need to undo swizzling in case of error anymore ✅

Reply:
update: method discovery (note: the only failure case is unfound method) and swizzling are separate now ✅